File: 547.friend-circles.go
func findCircleNum(M [][]int) int {
	row := len(M)
	col := len(M[0])

	var parent = make([]int, row)
	var size = make([]int, row)

	res := row

	uf_init(parent, size)

	for i := 0; i < row; i++ {
		for j := i+1; j < col; j++ {
			if M[i][j] == 1 {
				uf_add(parent, size, i, j, &res)
			}
		}
	}

	return res
}

func uf_init(parent []int, size []int) {
	for i := 0; i < len(parent); i++ {
		parent[i] = i
	}

	for i := 0; i < len(size); i++ {
		size[i] = 1
	}
}

func uf_add(parent []int, size []int, i int, j int, res *int) {
	px := uf_find(parent, i)
	py := uf_find(parent, j)

	if px == py {
		return;
	}

	if size[px] > size[py] {
		parent[py] = px
		size[px] += size[py]
	} else {
		parent[px] = py
		size[py] += size[px]
	}


	*res = *res - 1
}

func uf_find(parent []int, x int) int {
	for x != parent[x] {
		x = parent[x]
	}

	return x
}

File: 547.friend-circles.cpp
// CreateTime: 2019-12-23 18:04:29
class Solution {
public:
    
    vector<int> parent;

    int res;

    int find(int x) {
        if (x != parent[x]) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    void merge(int x, int y) {
        int px = find(x);
        int py = find(y);

        if (px == py) {
            return;
        }
        
        res -= 1;
        parent[py] = px;
    }

    int findCircleNum(vector<vector<int>>& M) {
        int len = M.size();

        res = len;
        parent.resize(len);

        for (int i = 0; i < len; i++) {
            parent[i] = i;
        }
        
        for (int i = 0; i < len; i++) {
            for (int j = 0; j < len; j++) {
                if (M[i][j] == 1) {
                    merge(i, j);
                }
            }
        }
        
        return res;
    }
};
